### PR TITLE
add sql_query() call to query the index

### PR DESCRIPTION
**Pull request by fedorov**
also:
* added corresponding test
* added loading of the index from the URL when local file is not found
* added detection of local s5cmd if not found where expected, and error handling

### CODE REVIEW

File: idc_index/index.py [hunk 1]
```diff
@@ -6,13 +6,17 @@
...
 import zipfile
+import duckdb
 
 
 class IDCClient:
     def __init__(self):
         current_dir = os.path.dirname(os.path.abspath(__file__))
         file_path = os.path.join(current_dir, 'idc_index.csv.zip')
-        self.index = pd.read_csv(file_path, dtype=str, encoding='utf-8')
+        if not os.path.exists(file_path):
```
Comment (fedorov) [inline idc_index/index.py:16]: @vkt1414 I added this and s5cmd detection below to facilitate local testing.

**Comment by fedorov** [comment]
To clarify, I believe replacing existing calls with duckdb is not a priority, can be done later. Most important is to enable users to execute their own queries that can utilize any combination of the columns in the index.

File: idc_index/index.py (hunk 1 above)
Comment (vkt1414) [reply idc_index/index.py:16]: I do check whether csv and s5cmd installed properly in the setup.py..
image
image

**Comment by vkt1414** [comment]
> To clarify, I believe replacing existing calls with duckdb is not a priority, can be done later. Most important is to enable users to execute their own queries that can utilize any combination of the columns in the index.

I think there is room for improvement regarding column names..as columns are named differently in the index.py and csv. Once we use same column names, this is a brilliant way to let users go wild, which is impossible with pandas

**Review by vkt1414 — CHANGES_REQUESTED** [review]

**Comment by fedorov** [comment]
> I donot know how this would be possible if pip installed the package properly..

pip is just one way to install the package. If I have the source code locally, and I want to run tests, I cannot do pip install.

**Comment by fedorov** [comment]
> I think there is room for improvement regarding column names..as columns are named differently in the index.py and csv.

Can you clarify? I am not sure I understand. I do not think we rename any columns in `self.index`.

**Comment by vkt1414** [comment]
> > I donot know how this would be possible if pip installed the package properly..
> 
> pip is just one way to install the package. If I have the source code locally, and I want to run tests, I cannot do pip install.

To test..all we need to do is python3 setup.py sdist and it will generate a tar.gz in dist folder..then we can install it with pip install path to the tar.gz file

**Comment by vkt1414** [comment]
> > I think there is room for improvement regarding column names..as columns are named differently in the index.py and csv.
> 
> Can you clarify? I am not sure I understand. I do not think we rename any columns in `self.index`.

what I meant was  for example, dicom_all seriesid is SeriesInstanceUID in get_dicom_series we use seriesInstanceUID argument, similarly for others too patientID vs PatientID in dicom_all. So when users use duckdb they will have to query against the original column names.(https://github.com/ImagingDataCommons/idc-index/assets/115020590/8e3d0fed-0612-4747-8633-92e6f8653a48)

**Comment by fedorov** [comment]
> To test..all we need to do is python3 setup.py sdist and it will generate a tar.gz in dist folder..then we can install it with pip install path to the tar.gz file

Well. Right now, to test, all I need to do is run this command from the root of the source code folder. This is so much more straightforward and simple!

```python -m unittest -vv tests/idcindex.py```

**Comment by fedorov** [comment]
> what I meant was for example, dicom_all seriesid is SeriesInstanceUID in get_dicom_series we use seriesInstanceUID argument, similarly for others too patientID vs PatientID in dicom_all. So when users use duckdb they will have to query against the original column names

We can make these consistent, but it doesn't bother me because those are names of the parameters to the functions, and they do not need to be one-to-one identical with the column names. But yes, we can rename those. We can do this in a next iteration while updating the functions to use `duckdb` at the same time. It will need to be coordinated with the Slicer extension, so it's a bit more work.

**Comment by vkt1414** [comment]
> > To test..all we need to do is python3 setup.py sdist and it will generate a tar.gz in dist folder..then we can install it with pip install path to the tar.gz file
> 
> Well. Right now, to test, all I need to do is run this command from the root of the source code folder. This is so much more straightforward and simple!
> 
> `python -m unittest -vv tests/idcindex.py`



> > To test..all we need to do is python3 setup.py sdist and it will generate a tar.gz in dist folder..then we can install it with pip install path to the tar.gz file
> 
> Well. Right now, to test, all I need to do is run this command from the root of the source code folder. This is so much more straightforward and simple!
> 
> `python -m unittest -vv tests/idcindex.py`

I do not know. Your call. In my opinion csv and s5cmd download should be handled by pip install alone and not the source code as you and steve convinced me in the first place.

**Comment by fedorov** [comment]
Yes, but it is and it will remain so - when using pip, those operations will not be invoked. But it's a fallback that can be used whenever the user did not install using pip. What is the alternative? Without this fallback, if the user legitimately uses unittest, for example, or direct package import from source code, there will be an irrecoverable error. When you develop the code and want to test it on the spot, pip install is not helpful.

**Comment by fedorov** [comment]
@pieper what is your opinion: right now, when the package is installed with pip install, it will fetch the csv index as a step of the installation. But if the package is used directly from source tree, for example, when running unit tests, there would be no preceding pip install, so I added a check that would load CSV from a URL whenever it is not available locally (which would not be the case after pip install). Do you have concerns about this approach?

**Comment by vkt1414** [comment]
> Yes, but it is and it will remain so - when using pip, those operations will not be invoked. But it's a fallback that can be used whenever the user did not install using pip. What is the alternative? Without this fallback, if the user legitimately uses unittest, for example, or direct package import from source code, there will be an irrecoverable error. When you develop the code and want to test it on the spot, pip install is not helpful.

Isn't it a normal practice to build first, and then perform tests on the build? sorry I do not have much experience with unit tests to know what's the best practice.

**Comment by fedorov** [comment]
I am not an expert in python testing either. It might be that the approach I use is not the best practice, but I found it very helpful and expedient in testing the changes I am making to the code before committing them.

**Comment by pieper** [comment]
> @pieper what is your opinion: right now, when the package is installed with pip install, it will fetch the csv index as a step of the installation. But if the package is used directly from source tree, for example, when running unit tests, there would be no preceding pip install, so I added a check that would load CSV from a URL whenever it is not available locally (which would not be the case after pip install). Do you have concerns about this approach?

I would think you want to offer some explicit api level of control about the versions of the csv file.  I.e. you get one during pip install because without it the package is not useful, but they should be versioned and you should have the option to check for updates and to use a specific version.  Your local tests should be pegged against a particular version anyway I think, or have a specific flag to check the "latest" csv so you will know if a regression is code or data related.

**Comment by fedorov** [comment]
I am going to merge this, because I need the SQL interface functionality for the tutorial, and there is no time right now, but I will create a separate issue to revisit this. I agree we should version the CSV and allow checking for updates, but I do not understand how this is related to the specific feature I added - it downloads the same CSV as pip install. And no matter what, the class constructor should be robust the absence of the CSV file or s5cmd. At the very least, there should be checks that those files exist - in the constructor. I do not think checks in setup.py are sufficient - the files could disappear after the installation!

**Comment by vkt1414** [comment]
I think that's a good idea to go ahead as it should not theoretically affect.

But I can't agree with files disappearing as installation is handled by pip and unless someone chooses to find where it is installed and removes them.

I'll work on how to build and test directly with setup.py and submit a PR sometime next week.

**Comment by fedorov** [comment]
> But I can't agree with files disappearing as installation is handled by pip and unless someone chooses to find where it is installed and removes them.

And I can't agree that the constructor does not check for the presence of csv index file and s5cmd binary! :-D

But I am sure we can find agreement in a discussion!

**Comment by vkt1414** [comment]
Absolutely. I'm sure we will find a better way to address as the repo evolves.

**Pull request merged**